File: python/jsonparser.py
'''
Author: Chen Cheng
Description: a class used to parse the json format string
'''

_ESCAPE_STR = ' \n\t\r'
_QUOTE_STR = '"\\/bfnrtu'
_ERR_MSG = 'Json format error'

class JsonParser:
    
    def __init__(self, ss=None):
        self._ss = ss
        self._index = 0
        self._line = 1
        self._pres = -1
        self._dd = {}

    def parse(self):
        try:
            self._escape()
            c = self._ss[self._index]
            self._index += 1
            if c == '{':
                self._dd = self._parse_object()
            elif c == '[':
                #self._dd[0] = self._parse_array(i)
                self._dd = self._parse_array()
            else:
                raise JsonParseError(_ERR_MSG, self._ss, self._index)
            self._escape()
            if self._index < len(self._ss):
                raise JsonParseError(_ERR_MSG + '--extra data', self._ss, self._index)
            return self._dd
        except JsonParseError as e:
            e.err_msg()

    def _parse_object(self):
        try:
            dd = {}
            self._escape()
            # empty object {}
            if self._ss[self._index] == '}':
                self._index += 1
                return dd
            elif self._ss[self._index] != '"':
                raise JsonParseError(_ERR_MSG, self._ss, self._index)
            
            self._index += 1
            while True:
                key = self._parse_string()
                self._escape()
                if self._ss[self._index] != ':':
                    raise JsonParseError(_ERR_MSG, self._ss, self._index)
                self._index = self._index + 1
                self._escape()
                dd[key] = self._parse_value()
                self._escape()
                if self._ss[self._index] == '}':
                    self._index = self._index + 1
                    break
                elif self._ss[self._index] == ',':
                    self._index = self._index + 1
                    self._escape()

                if self._ss[self._index] != '"':
                    raise JsonParseError(_ERR_MSG, self._ss, self._index)
                self._index = self._index + 1
            return dd
        except JsonParseError as e:
            e.err_msg()

    def _parse_array(self):
        ll = []
        self._escape()
        # empty array []
        if self._ss[self._index] == ']':
            self._index += 1
            return ll
        while True:
            v = self._parse_value()
            ll.append(v)
            self._escape()
            if self._ss[self._index] == ',':
                self._index = self._index + 1
                self._escape()
            elif self._ss[self._index] == ']':
                self._index = self._index + 1
                return ll
            else:
                raise JsonParseError(_ERR_MSG, self._ss, self._index)

    def _parse_string(self):
        begin = end = self._index
        try:
            while self._ss[end] != '"':
                if self._ss[end] == '\\':
                    end = end + 1
                    if self._ss[end] not in _QUOTE_STR:
                        raise JsonParseError(_ERR_MSG, self._ss, self._index)
                end = end + 1
        except IndexError:
            raise JsonParseError(_ERR_MSG, self._ss, self._index)
        self._index = end + 1
        return self._ss[begin:end]
    
    def _parse_number(self):
        begin = end = self._index
        end_str = _ESCAPE_STR + ',}]'
        try:
            while self._ss[end] not in end_str:
                end = end + 1
            ns = self._ss[begin:end]
            if '.' in ns or 'e' in ns or 'E' in ns or ns == 'int' or ns == '-inf':
                n = float(ns)
            else:
                n = int(ns)
            self._index = end
            return n
        except IndexError:
            raise JsonParseError(_ERR_MSG, self._ss, self._index)

    def _parse_value(self):
        v = None
        c = self._ss[self._index]
        if c == '{':
            self._index = self._index + 1
            v = self._parse_object()
        elif c == '[':
            self._index = self._index + 1
            v = self._parse_array()
        elif c == '"':
            self._index = self._index + 1
            v = self._parse_string()
        elif c == 'n' and self._ss[self._index: self._index + 4] == 'null':
            self._index = self._index + 4
            v = None
        elif c == 't' and self._ss[self._index: self._index + 4] == 'true':
            self._index = self._index + 4
            v = True
        elif c == 'f' and self._ss[self._index: self._index + 5] == 'false':
            self._index = self._index + 5
            v = False
        else:
            v = self._parse_number()
        return v

    def _escape(self):
        while self._index < len(self._ss) and self._ss[self._index] in _ESCAPE_STR:
            self._index = self._index + 1

class JsonParseError(ValueError):
    def __init__(self, msg, ss, index):
        self._msg = msg
        self._ss = ss
        self._index = index
        self._count_line_col()
        #self.err_msg()
        #assert(0)

    def err_msg(self):
        print '%s: line %d column %d char %s' % (self._msg, self._line, self._col, self._ss[self._index])

    def _count_line_col(self):
        self._line = self._ss.count('\n', 0, self._index) + 1
        if self._line == 1:
            self._col = self._index
        else:
            self._col = self._index - self._ss.rindex('\n', 0, self._index)

File: lua/jsonparser.lua
--[[
--Author: Chen Cheng
--Description:
--]]

local ESCAPE_STR = ' \n\t\r'
local QUOTE_STR = '"\\/bfnrtu'
local ERR_MSG = 'json parse error'
local parse_array
local parse_object

local function null()
    return null
end

local function get(ss, index)
    return string.sub(ss, index, index)
end

local function parse_error(msg, ss, index)
    local line = 0
    local col = 0
    local ts = string.sub(ss, 1, index)
    _, line = string.gsub(ts, '\n', '')
    line = line + 1
    if line == 1 then
        col = index
    else
        local rts = string.reverse(ts)
        local pos = 0
        pos = string.find(rts, '\n')
        col = pos - 1
    end
    print(string.format('%s: line %d colunm %d char %s', msg, line, col, get(ss, index)))
    assert(nil)
end

local function escape(ss, index)
    while index <= string.len(ss) and 
        string.find(ESCAPE_STR, get(ss, index), nil, true) do
        index = index + 1
    end
    return index
end

local function parse_number(ss, index)
    local b = index
    local e = index
    end_str = ESCAPE_STR .. ',}]'
    while string.find(end_str, get(ss, e), nil, true) == nil do
        e = e + 1
    end
    local ns = string.sub(ss, b, e - 1)
    n = tonumber(ns)
    index = e
    if n == nil then
        parse_error(ERR_MSG, ss, index)
    end
    return n, index
end

local function parse_string(ss, index)
    local b = index
    local e = index
    while get(ss, e) ~= '"' do
        if get(ss, e) == '\\' then
            e = e + 1
            if string.find(QUOTE_STR, get(ss, e), nil, true) == nil then
                parse_error(ERR_MSG, ss, index)
            end
        end
        e = e + 1
    end
    index = e + 1
    return string.sub(ss, b, e - 1), index
end

local function parse_value(ss, index)
    local v = nil
    local c = get(ss, index)
    if c == '{' then
        index = index + 1
        v, index = parse_object(ss, index)
    elseif c == '[' then
        index = index + 1
        v, index = parse_array(ss, index)
    elseif c == '"' then
        index = index + 1
        v, index = parse_string(ss, index)
    elseif c == 'n' and string.sub(ss, index, index + 3) == 'null' then
        index = index + 4
        v = null
    elseif c == 't' and string.sub(ss, index, index + 3) == 'true' then
        index = index + 4
        v = true
    elseif c == 'f' and string.sub(ss, index, index + 4) == 'false' then
        index = index + 5
        v = false
    else
        v, index = parse_number(ss, index)
    end
    return v, index
end

parse_array = function (ss, index)
    local ll = {}
    ll[0] = 'array'
    index = escape(ss, index)
    if get(ss, index) == ']' then
        index = index + 1
        return ll, index
    end
    local pos = 1
    while true do
        ll[pos], index = parse_value(ss, index)
        pos = pos + 1
        index = escape(ss, index)
        if get(ss, index) == ',' then
            index = index + 1
            index = escape(ss, index)
        elseif get(ss, index) == ']' then
            index = index + 1
            break
        else
            parse_error(ERR_MSG, ss, index)
        end
    end
    return ll, index
end

parse_object = function (ss, index)
    local tt = {}
    tt[0] = 'object'
    index = escape(ss, index)
    if get(ss, index) == '}' then
        index = index + 1
        return tt, index
    elseif get(ss, index) ~= '"' then
        parse_error(ERR_MSG, ss, index)
    end
    index = index + 1
    local key = ''
    while true do
        key, index = parse_string(ss, index)
        index = escape(ss, index)
        if get(ss, index) ~= ':' then
            parse_error(ERR_MSG, ss, index)
        end
        index = index + 1
        index = escape(ss, index)
        tt[key], index = parse_value(ss, index)
        index = escape(ss, index)
        if get(ss, index) == '}' then
            index = index + 1
            break
        elseif get(ss, index) == ',' then
            index = index + 1
            index = escape(ss, index)
        end

        if get(ss, index) ~= '"' then
            parse_error(ERR_MSG, ss, index)
        end
        index = index + 1
    end
    return tt, index
end

local function parse(ss)
    local index = 1
    local tt = {}
    index = escape(ss, index)
    local c = get(ss, index)
    index = index + 1
    if c == '{' then
        tt, index = parse_object(ss, index)
    elseif c == '[' then
        tt, index = parse_array(ss, index)
    end 
    index = escape(ss, index)
    if index <= string.len(ss) then
        parse_error(ERR_MSG .. '--extra data', ss, index)
    end
    return tt
end

local jsonparser = {
    parse = parse,
    null = null
}

return jsonparser
